r2/r2/lib/providers/search/solrsearch.py

```diff
@@ -0,0 +1,812 @@
+                else:
+                    raise SearchHTTPError(response.status,
+                                               response.reason,
+                                               response.read())
+        finally:
+            connection.close()
+        return responses
+
+
+def chunk_xml(xml, depth=0):
+    '''Chunk POST data into pieces that are smaller than the 20 MB limit.
+    
+    Ideally, this never happens (if chunking is necessary, would be better
+    to avoid xml'ifying before testing content_length)'''
+    data = etree.tostring(xml)
+    root = xml.findall('add') and 'add' or 'delete'
+    content_length = len(data)
+    if content_length < _CHUNK_SIZE:
+        yield data
+    else:
+        depth += 1
+        print "WARNING: Chunking (depth=%s)" % depth
+        half = len(xml) / 2
+        left_half = xml # for ease of reading
+        right_half = etree.Element(root)
+        # etree magic simultaneously removes the elements from one tree
+        # when they are appended to a different tree
+        right_half.append(xml[half:])
+        for chunk in chunk_xml(left_half, depth=depth):
+            yield chunk
+        for chunk in chunk_xml(right_half, depth=depth):
+            yield chunk
+
+
+@g.stats.amqp_processor('solrsearch_q')
+def _run_changed(msgs, chan):
+    '''Consume the cloudsearch_changes queue, and print reporting information
+    on how long it took and how many remain
+    
+    '''
+    start = datetime.now(g.tz)
+
+    changed = [pickle.loads(msg.body) for msg in msgs]
+
+    link_fns = SolrLinkUploader.desired_fullnames(changed)
+    sr_fns = SolrSubredditUploader.desired_fullnames(changed)
+
+    link_uploader = SolrLinkUploader(g.solr_doc_host, fullnames=link_fns)
+    subreddit_uploader = SolrSubredditUploader(g.solr_subreddit_doc_host,
+                                           fullnames=sr_fns)
+
+    link_time = link_uploader.inject()
+    subreddit_time = subreddit_uploader.inject()
+    solrsearch_time = link_time + subreddit_time
+
+    totaltime = (datetime.now(g.tz) - start).total_seconds()
+
+    print ("%s: %d messages in %.2fs seconds (%.2fs secs waiting on "
+           "solr); %d duplicates, %s remaining)" %
+           (start, len(changed), totaltime, solrsearch_time,
+            len(changed) - len(link_fns | sr_fns),
+            msgs[-1].delivery_info.get('message_count', 'unknown')))
+
+
+class SolrLinkUploader(SolrSearchUploader):
+    types = (Link,)
+
+    def __init__(self, solr_host=None, solr_port=None, fullnames=None):
+        super(SolrLinkUploader, self).__init__(fullnames=fullnames)
+        self.accounts = {}
+        self.srs = {}
+
+    def fields(self, thing):
+        '''Return fields relevant to a Link search index'''
+        account = self.accounts[thing.author_id]
+        sr = self.srs[thing.sr_id]
+        return LinkFields(thing, account, sr).fields()
+
+    def batch_lookups(self):
+        super(SolrLinkUploader, self).batch_lookups()
```

Comment:
There's no `batch_lookups()` function on the super class, so this line is failing.